File: ios/InfoNeigeObjC/InfoNeigeObjC/ViewController.m
//
//  ViewController.m
//  InfoNeigeObjC
//
//  Created by Van Du Tran on 2014-07-29.
//  Copyright (c) 2014 Van Du Tran. All rights reserved.
//

#import "ViewController.h"
@import CoreLocation;
@import MapKit;

@interface ViewController () <CLLocationManagerDelegate, MKMapViewDelegate>
@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (strong, nonatomic) CLLocationManager *locationManager;
@end

@implementation ViewController
            
- (void)viewDidLoad {
    [super viewDidLoad];

    self.locationManager = [[CLLocationManager alloc] init];
    [self.locationManager requestWhenInUseAuthorization];
    self.mapView.userTrackingMode = MKUserTrackingModeFollow;
    self.mapView.delegate = self;
    
    [self countriesOverlays];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)countriesOverlays {
    NSString *fileName = [[NSBundle mainBundle] pathForResource:@"cote" ofType:@"json"];
    NSData *overlayData = [NSData dataWithContentsOfFile:fileName];
    
    NSArray *sides = [[NSJSONSerialization JSONObjectWithData:overlayData options:NSJSONReadingAllowFragments error:nil] objectForKey:@"features"];
    
//    NSMutableArray *overlays = [NSMutableArray array];
    
    for (NSDictionary *side in sides) {
        
        NSDictionary *geometry = side[@"geometry"];
        if ([geometry isKindOfClass:[NSNull class]]) {
            continue;
        }
        
        if ([geometry[@"type"] isEqualToString:@"LineString"]) {
            NSArray *coordinates = geometry[@"coordinates"];
            
            CLLocationCoordinate2D *pointArr = malloc(sizeof(CLLocationCoordinate2D) * [coordinates count]);
            
            int idx = 0;
            for (NSArray *coordinate in coordinates) {
            
                pointArr[idx] = CLLocationCoordinate2DMake([coordinate[0] doubleValue], [coordinate[1] doubleValue]);
                idx++;
            }
            
            MKPolyline *path = [MKPolyline polylineWithCoordinates:pointArr count:[coordinates count]];
            path.title = side[@"properties"][@"NOM_VOIE"];
            [self.mapView addOverlay:path level:MKOverlayLevelAboveRoads];
//            [overlays addObject:path];
            
            
            
            
//            MKPolygon *polygon = [HHLViewController overlaysFromPolygons:geometry[@"coordinates"] id:country[@"properties"][@"name"]];
//            if (polygon) {
//                [overlays addObject:polygon];
//            }
            
        }
//        } else if ([geometry[@"type"] isEqualToString:@"MultiPolygon"]){
//            for (NSArray *polygonData in geometry[@"coordinates"]) {
//                MKPolygon *polygon = [HHLViewController overlaysFromPolygons:polygonData id:country[@"properties"][@"name"]];
//                if (polygon) {
//                    [overlays addObject:polygon];
//                }
//            }
//        } else {
//            NSLog(@"Unsupported type: %@", geometry[@"type"]);
//        }
    }
    
//    [self.mapView addOverlay:overlays level:MKOverlayLevelAboveRoads];

    
}

- (MKOverlayRenderer *)mapView:(MKMapView *)mapView rendererForOverlay:(id<MKOverlay>)overlay
{
    if ([overlay isKindOfClass:[MKPolyline class]]) {
        MKPolylineRenderer *renderer = [[MKPolylineRenderer alloc] initWithOverlay:overlay];
        renderer.strokeColor = [UIColor colorWithRed:74.0/255.0 green:144.0/255.0 blue:226.0/255.0 alpha:1.0f];
        renderer.lineWidth = 6.0f;
        return renderer;
    }
    
    return nil;
}

@end

File: ios/InfoNeige/InfoNeige/ViewController.swift
//
//  ViewController.swift
//  InfoNeige
//
//  Created by Van Du Tran on 2014-07-27.
//
//

import CoreLocation
import MapKit
import UIKit

class ViewController: UIViewController, CLLocationManagerDelegate, MKMapViewDelegate {
    @IBOutlet weak var mapView: MKMapView!
    let locationManager: CLLocationManager = CLLocationManager ()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        locationManager.requestWhenInUseAuthorization()
        self.mapView.userTrackingMode = MKUserTrackingMode.Follow
        self.mapView.delegate = self
    }
    
    override func viewWillAppear(animated: Bool)
    {
        super.viewWillAppear(animated)
        
//        self.streetSidesOverlays()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    
    /*
    func streetSidesOverlays() {
        let fileName = NSBundle.mainBundle().pathForResource("cote", ofType: "json")
        var error: NSError?
        let overlayData = NSData.dataWithContentsOfFile(fileName, options: NSDataReadingOptions.DataReadingMapped, error: &error)
        let streetSides: NSDictionary = NSJSONSerialization.JSONObjectWithData(overlayData, options: NSJSONReadingOptions.AllowFragments, error: nil) as NSDictionary
        let sides: NSArray = streetSides.objectForKey("features") as NSArray
        for side in sides {
            if let geometry = side["geometry"] as? NSDictionary {
                
                // a one line string
                if let type = geometry["type"] as? NSString {
                    if type.isEqualToString("LineString") {
                        let coordinates: NSArray = geometry["coordinates"] as NSArray
                        
                        var oneSegment = [CLLocationCoordinate2D]()
                        
                        for coordinate in coordinates {
                            let numberOfCoordinates: UInt8 = UInt8(coordinates.count.asUnsigned())
                            var latitude: CLLocationDegrees = coordinate[0] as CLLocationDegrees
                            var longitude: CLLocationDegrees = coordinate[1] as CLLocationDegrees
                            var onePoint:CLLocationCoordinate2D = CLLocationCoordinate2DMake(latitude, longitude)
                            
                            oneSegment += onePoint
                        }
                        
                        var path = MKPolyline(coordinates: &oneSegment, count: coordinates.count)
//                        println("\(oneSegment)")
                        
                        self.mapView.addOverlay(path, level: MKOverlayLevel.AboveRoads)
//                        println(self.mapView.overlays)
//                        self.mapView.addOverlays([path], level: MKOverlayLevel.AboveRoads)
                    }
                }
                else {
                    continue;
                }
                
            }
            else {
                continue;
            }
        }
    }
*/
    
    func mapView(mapView: MKMapView!, didUpdateUserLocation userLocation: MKUserLocation!) {
        println("didUpdateUserLocation")
    }
    
    /*
    func mapView(mapView: MKMapView!, rendererForOverlay overlay: MKOverlay!) -> MKOverlayRenderer! {
        println("rendererForOverlay")
        if overlay is MKPolyline {
            var renderer = MKPolylineRenderer(overlay: overlay)
            renderer.strokeColor = UIColor(red: 74.0/255.0, green: 144.0/255.0, blue: 226.0/255/0, alpha: 1.0)
            renderer.lineWidth = 6.0
            return renderer
        }
        
        return nil
    }
*/
}

